File: go/binary-search/binary_search.go
package binarysearch

import "fmt"

// SearchInts performs a binary search for key in list and returns the index if
// key is found. Returns `-1` if key is not in list.
func SearchInts(list []int, key int) (index int) {
	return binarySearch(list, key, 0, len(list))
}

func binarySearch(list []int, key int, start int, end int) (index int) {
	if len(list) == 0 {
		return -1
	}
	if end < start {
		return -1
	}
	if start >= len(list) {
		return -1
	}
	if start == end {
		if key == list[start] {
			return start
		} else {
			return -1
		}
	}
	middle := ((end - start) / 2) + start
	fmt.Printf("binarySearch(%v, %v, %v, %v)\n", list, key, start, end)
	if key == list[middle] {
		return middle
	} else if key > list[middle] {
		return binarySearch(list, key, middle+1, end)
	} else if key < list[middle] {
		return binarySearch(list, key, start, middle)
	}
	fmt.Printf("unreachable\n")
	return -1
}

File: typescript/binary-search/binary-search.ts
/**
 * Find performs a binary search for `item` in `array` and returns its index if
 * found. Array must be provided in sorted order. `find` throws an Error if
 * value if nost in the array.
 */
export function find(array: number[], item: number): number {
  if (array.length === 0) {
    throw new Error('Value not in array')
  }

  return binarySearch(array, item, 0, array.length - 1)
}

function binarySearch(array: number[], item: number, left: number, right: number): number {
  if(right < left) {
    console.log(`right: ${right}, left: ${left}`)
    throw new Error('Value not in array')
  }
  if (right == left && array[right] !== item) {
    throw new Error('Value not in array')
  }
  if (right == left && array[right] === item) {
    return right
  }

  const middle = Math.floor((right - left) / 2) + left
  const middleVal = array[middle]
  console.log(`right: ${right}, left: ${left}, middle: ${middle}, middleVal ${middleVal}`)
  if (item === middleVal) {
    return middle
  }
  if (item > middleVal) {
    return binarySearch(array, item, middle + 1, right)
  }
  if (item < middleVal) {
    return binarySearch(array, item, left, middle - 1)
  }
  throw new Error("unreachable")
}
